akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -0,0 +1,643 @@
+            "removing from MergeHub now", ex)
+        }
+
+        private def onDemand(moreDemand: Long): Unit = {
+          if (moreDemand == MergeHub.Cancel) completeStage()
+          else {
+            demand += moreDemand
+            if (!hasBeenPulled(in)) pullWithDemand()
+          }
+        }
+
+        setHandler(in, this)
+      }
+
+    }
+
+    (logic, Sink.fromGraph(sink))
+  }
+}
+
+/**
+ * A BroadcastHub is a special streaming hub that is able to broadcast streamed elements to a dynamic set of consumers.
+ * It consissts of two parts, a [[Sink]] and a [[Source]]. The [[Sink]] broadcasts elements from a producer to the
+ * actually live consumers it has. Once the producer has been materialized, the [[Sink]] it feeds into returns a
+ * materialized value which is the corresponding [[Source]]. This [[Source]] can be materialized arbitrary many times,
+ * where weach of the new materializations will receive their elements from the original [[Sink]].
+ */
+object BroadcastHub {
+
+  /**
+   * Creates a [[Sink]] that receives elements from its upstream producer and broadcasts them to a dynamic set
+   * of consumers. After the [[Sink]] returned by this method is materialized, it returns a [[Source]] as materialized
+   * value. This [[Source]] can be materialized arbitrary many times and each materialization will receive the
+   * broadcast elements form the ofiginal [[Sink]].
+   *
+   * Every new materialization of the [[Sink]] results in a new, independent hub, which materializes to its own
+   * [[Source]] for consuming the [[Sink]] of that materialization.
+   *
+   * If the original [[Sink]] is failed, then the failure is immediately propagated to all of its materialized
+   * [[Source]]s (possibly jumping over already buffered elements). If the original [[Sink]] is completed, then
+   * all corresponding [[Source]]s are completed. Both failure and normal completion is "remembered" and later
+   * materializations of the [[Source]] will see the same (failure or completion) state. [[Source]]s that are
+   * cancelled are simply removed from the dynamic set of consumers.
+   *
+   * @param clazz Type of elements this hub emits and consumes
+   * @param bufferSize Buffer size used by the producer. Gives an upper bound on how "far" from each other two
+   *                   concurrent consumers can be in terms of element. If this buffer is full, the producer
+   *                   is backpressured.
+   */
+  def sink[T](bufferSize: Int): Sink[T, Source[T, NotUsed]] = Sink.fromGraph(new BroadcastHub[T](bufferSize))
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMaterializedValue[SinkShape[T], Source[T, NotUsed]] {
+  require(bufferSize > 0, "Buffer size must be positive")
+  require(bufferSize < 4096, "Buffer size larger then 4095 is not allowed")
+  // TODO: Check for power of two
```

Comment:
Nope, I haven't pushed those yet. Left it yesterday as it is. Apart from TODOs the rest is free game :)